tools/encryption/pkg/encryption-helper.go

```diff
@@ -0,0 +1,263 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package pkg
+
+import (
+	"bytes"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"github.com/golang/protobuf/proto"
+	"github.com/google/tink/go/aead"
+	"github.com/google/tink/go/core/registry"
+	"github.com/google/tink/go/hybrid"
+	"github.com/google/tink/go/insecurecleartextkeyset"
+	"github.com/google/tink/go/keyset"
+	gcmpb "github.com/google/tink/proto/aes_gcm_go_proto"
+	tinkpb "github.com/google/tink/proto/tink_go_proto"
+	"golang.org/x/net/html"
+	"golang.org/x/net/html/atom"
+	"io"
+	"net/http"
+	"strings"
+)
+
+// Helper functions for the SwG Encryption Script.
+
+const aesGcmKeyUrl string = "type.googleapis.com/google.crypto.tink.AesGcmKey"
+const aesGcmKeySize uint32 = 16
+
+// Public function to generate an encrypted HTML document given the original.
+func GenerateEncryptedDocument(htmlStr, accessRequirement string, pubKey *tinkpb.Keyset) (string, error) {
+	km, err := registry.GetKeyManager(aesGcmKeyUrl)
+	if err != nil {
+		return "", err
+	}
+	key, err := generateNewAesGcmKey(km)
+	if err != nil {
+		return "", err
+	}
+	ks := createAesGcmKeyset(key)
+	ksEnc, err := proto.Marshal(&ks)
+	if err != nil {
+		return "", err
+	}
+	kh, err := insecurecleartextkeyset.Read(&keyset.MemReaderWriter{Keyset: &ks})
+	if err != nil {
+		return "", err
+	}
+	parsedHtml, err := html.Parse(strings.NewReader(htmlStr))
+	if err != nil {
+		return "", err
+	}
+	encryptedSections := getAllEncryptedSections(parsedHtml)
+	if err = encryptAllSections(parsedHtml, encryptedSections, kh); err != nil {
+		return "", err
+	}
+	encryptedKey, err := encryptDocumentKey(base64.StdEncoding.EncodeToString(ksEnc), accessRequirement, *pubKey)
+	if err != nil {
+		return "", err
+	}
+	if err = addEncryptedDocumentKeyToHead(encryptedKey, parsedHtml); err != nil {
+		return "", err
+	}
+	return renderNode(parsedHtml, false), nil
+}
+
+// Retrieves Google's public key from the given URL.
+func RetrieveGooglePublicKey(publicKeyUrl string) (tinkpb.Keyset, error) {
+	resp, err := http.Get(publicKeyUrl)
+	if err != nil {
+		return tinkpb.Keyset{}, err
+	}
+	r := keyset.NewJSONReader(resp.Body)
+	ks, err := r.Read()
+	if err != nil {
+		return tinkpb.Keyset{}, err
+	}
+	return *ks, nil
+}
+
+// Generates a new AES-GCM key.
+func generateNewAesGcmKey(km registry.KeyManager) ([]byte, error) {
+	p, err := proto.Marshal(&gcmpb.AesGcmKeyFormat{KeySize: aesGcmKeySize})
+	if err != nil {
+		return nil, err
+	}
+	m, err := km.NewKey(p)
+	if err != nil {
+		return nil, err
+	}
+	return proto.Marshal(m)
+}
+
+// Creates an AES-GCM Keyset using the input key.
```

Comment:
Why does this need to be a keyset as opposed to a single key?

Reply:
This is due to a specification by the Tink team. I'll add you into the email discussion.

Reply:
Any chance you could explicate in comments some examples of what serialized form of each part looks like?

Reply:
Added an example proto in the comments of the function.